hcl2template/addrs/plugin.go

```diff
@@ -13,21 +17,53 @@ import (
 
 // Plugin encapsulates a single plugin type.
 type Plugin struct {
-	Hostname  string
-	Namespace string
-	Type      string
+	Source string
 }
 
-func (p Plugin) RealRelativePath() string {
-	return p.Namespace + "/packer-plugin-" + p.Type
+// Parts returns the list of components of the source URL, starting with the
+// host, and ending with the name of the plugin (non-simplified).
+//
+// This will correspond more or less to the filesystem hierarchy where
+// the plugin is installed.
+func (p Plugin) Parts() []string {
+	var retParts []string
+
+	// This should not happen, we're supposed to check that the source isn't
+	// empty when creating the plugin, but let's be careful here and not
+	// misbehave.
+	if p.Source == "" {
+		log.Printf("[ERROR] - Plugin.Parts: the source was empty, this should not happen in normal usage, and is likely a Packer bug.")
+		return retParts
+	}
+
+	dir, part := path.Split(strings.TrimRight(p.Source, "/"))
+	for dir != "" {
+		retParts = append(retParts, part)
+		dir, part = path.Split(strings.TrimRight(dir, "/"))
+	}
+	retParts = append(retParts, part)
```

Comment:
I don't understand why we are doing this. Can you explain.

Reply:
The last `append` you mean? It's because the last `split` above returns an empty `dir` and a non-empty `part`, so we leave the loop, therefore that last `part` is not added to the slice, hence why we call it once more after that.

To illustrate:

```
p.Source = "github.com/hashicorp/amazon"
# Loop 1
dir = "github.com/hashicorp"
part = "amazon"
retParts = ["amazon"]
# Loop 2
dir = "github.com"
part = "hashicorp"
retParts = ["amazon", "hashicorp"]
# Loop 3
dir = ""
part = "github.com"
retParts = ["amazon", "hashicorp"]
# Out of loop
retParts = ["amazon", "hashicorp", "github.com"]]
```

Reply:
Sorry I should've been more specific with my questions. I was trying to figure out why you split the string this way. 

Looking at it again I suspect you want to split after each path separator '/' but since you don't know the count you keep calling pat.Split until you have all parts. Is that correct? 

If so, we can achieve this differently without needing the final append or slice.Reverse. 

```go
	source := strings.TrimRight(p.Source, "/")
	parts := strings.FieldsFunc(source, func(c rune) bool {
		return c == '/'
	})
```

Reply:
Strings may do the trick, unless we have an escaped `/` in the middle of the path (though I guess in valid URLs they should be URL-encoded?).
I went with `path.Split` since it's URL-centric, and I trust it will do the splitting according to URL-rules (all of which I don't know).

But you are right: conceptually I went with this approach as we don't know how many parts there are in that URL, and I didn't find a convenience method to explode the URL in its parts.

Reply:
Note: after talking out-of-band and looking into the actual implementation, `path.Split` doesn't do much more than a `strings.Split` so we can go with this implementation instead.